File: app/src/main/java/com/ndmsystems/coala/crypto/Aead.java
package com.ndmsystems.coala.crypto;

import com.ndmsystems.coala.helpers.Hex;
import com.ndmsystems.infrastructure.logging.LogHelper;

import java.nio.ByteBuffer;


public class Aead {
    public final byte[] peerKey;
    public final byte[] myKey;
    public final byte[] peerIV;
    public final byte[] myIV;
    private final AesGcm encryptor;
    private final AesGcm decryptor;

    public Aead(byte[] peerKey, byte[] myKey, byte[] peerIV, byte[] myIV) {
        LogHelper.v("Aead peerKey=" + Hex.encodeHexString(peerKey) + ", length: " + peerKey.length);
        LogHelper.v("Aead myKey=" + Hex.encodeHexString(myKey) + ", length: " + myKey.length);
        LogHelper.v("Aead peerIV=" + Hex.encodeHexString(peerIV) + ", length: " + peerIV.length);
        LogHelper.v("Aead myIV=" + Hex.encodeHexString(myIV) + ", length: " + myIV.length);
        this.peerKey = peerKey;
        this.myKey = myKey;
        this.peerIV = peerIV;
        this.myIV = myIV;
        encryptor = new AesGcm(this.myKey);
        decryptor = new AesGcm(this.peerKey);
    }


    public byte[] decrypt(byte[] cipherText, int counter, byte[] associatedData){
        try {
            LogHelper.v("Open, nonce: " + Hex.encodeHexString(makeNonce(peerIV, counter)));
            return decryptor.open(cipherText, makeNonce(peerIV, counter), associatedData);
        } catch (Exception e) {
            e.printStackTrace();
            LogHelper.e("Error then decrypt: " + e.getMessage() + ", nonce: " + Hex.encodeHexString(makeNonce(peerIV, counter)));
            return null;
        }
    }

    public byte[] encrypt(byte[] plainText, int counter, byte[] associatedData){
        try {
            return encryptor.seal(plainText, makeNonce(myIV, counter), associatedData);
        } catch (Exception e) {
            e.printStackTrace();
            LogHelper.e("Error then encrypt: " + e.getMessage());
            return null;
        }
    }

    private byte[] makeNonce(byte[] iv, int counter) {
        ByteBuffer nonce = ByteBuffer.allocate(12);
        nonce.put(iv).put((byte)(counter & 0xFF)).put((byte)((counter >> 8) & 0xFF));
        return nonce.array();
    }
}

File: app/src/main/java/com/ndmsystems/coala/crypto/Aead.kt
package com.ndmsystems.coala.crypto

import com.ndmsystems.coala.helpers.Hex
import com.ndmsystems.coala.helpers.logging.LogHelper
import com.ndmsystems.coala.helpers.logging.LogKeys
import java.nio.ByteBuffer

class Aead(peerKey: ByteArray, myKey: ByteArray, peerIV: ByteArray, myIV: ByteArray) {
    private val peerKey: ByteArray
    private val myKey: ByteArray
    private val peerIV: ByteArray
    private val myIV: ByteArray
    private val encryptor: AesGcm
    private val decryptor: AesGcm

    init {
        // The names matter: LogSanitizer blanks these four by key before anything is uploaded.
        // Logcat still shows them - ILogger.log's default folds the fields back into the text -
        // so local debugging keeps the material and the wire never sees it.
        LogHelper.v(
            "Aead keys",
            mapOf(
                "peer_key" to Hex.encodeHexString(peerKey),
                "peer_key_length" to peerKey.size,
                "my_key" to Hex.encodeHexString(myKey),
                "my_key_length" to myKey.size,
                "peer_iv" to Hex.encodeHexString(peerIV),
                "peer_iv_length" to peerIV.size,
                "my_iv" to Hex.encodeHexString(myIV),
                "my_iv_length" to myIV.size
            )
        )
        this.peerKey = peerKey
        this.myKey = myKey
        this.peerIV = peerIV
        this.myIV = myIV
        encryptor = AesGcm(this.myKey)
        decryptor = AesGcm(this.peerKey)
    }

    fun decrypt(cipherText: ByteArray, counter: Int, associatedData: ByteArray?): ByteArray? {
        return try {
            decryptor.open(cipherText, makeNonce(peerIV, counter), associatedData)
        } catch (e: Exception) {
            e.printStackTrace()
            LogHelper.e(
                "Error then decrypt",
                mapOf(
                    LogKeys.ERROR to e.message,
                    "nonce" to Hex.encodeHexString(makeNonce(peerIV, counter))
                )
            )
            null
        }
    }

    override fun toString(): String {
        return Hex.encodeHexString(peerKey) + Hex.encodeHexString(myKey) + Hex.encodeHexString(peerIV) + Hex.encodeHexString(myIV)
    }

    fun encrypt(plainText: ByteArray, counter: Int, associatedData: ByteArray?): ByteArray? {
        return try {
            encryptor.seal(plainText, makeNonce(myIV, counter), associatedData)
        } catch (e: Exception) {
            e.printStackTrace()
            LogHelper.e(
                "Error then encrypt",
                mapOf(LogKeys.ERROR_TYPE to e.javaClass.name, LogKeys.ERROR to e.message)
            )
            null
        }
    }

    private fun makeNonce(iv: ByteArray, counter: Int): ByteArray {
        val nonce = ByteBuffer.allocate(12)
        nonce.put(iv).put((counter and 0xFF).toByte()).put((counter shr 8 and 0xFF).toByte())
        return nonce.array()
    }
}